### PR TITLE
Increase PackageBuilder unit test coverage

### PR DESCRIPTION
**Pull request by aciidb0mb3r**
This almost fully covers package builder by unit tests except the code hack for injecting dependencies of swiftpm

### CODE REVIEW

File: Tests/PackageLoading/ConventionTests.swift [hunk 1]
```diff
@@ -440,6 +440,95 @@ class ConventionTests: XCTestCase {
         }
     }
 
+    func testValidSources() throws {
+        var fs = InMemoryFileSystem()
+        try fs.createEmptyFiles("/main.swift",
+                                "/noExtension",
+                                "/Package.swift",
+                                "/.git/anchor",
+                                "/.xcodeproj/anchor",
```
Comment (ddunbar) [inline Tests/PackageLoading/ConventionTests.swift]: Should this be a `.swift` to verify we ignore such files in ignored directories?

File: Tests/PackageLoading/ConventionTests.swift [hunk 2]
```diff
@@ -440,6 +440,95 @@ class ConventionTests: XCTestCase {
         }
     }
 
+    func testValidSources() throws {
+        var fs = InMemoryFileSystem()
+        try fs.createEmptyFiles("/main.swift",
```
Comment (ddunbar) [inline Tests/PackageLoading/ConventionTests.swift]: Completely off topic, but I wonder whether we should really support this particular convention... it seems pretty rarely used, and it is confusing to have the Swift files adjacent to the manifest.

File: Tests/PackageLoading/ConventionTests.swift [hunk 3]
```diff
@@ -440,6 +440,95 @@ class ConventionTests: XCTestCase {
...
+        let name = "pkg"
+        PackageBuilderTester(name, in: fs) { result in
+            result.checkModule(name) { moduleResult in
+                moduleResult.check(type: .executable, isTest: false)
+                moduleResult.checkSources(root: "/", paths: "main.swift")
+            }
+        }
+    }
+
+    func testCustomTargetDependencies() throws {
```
Comment (ddunbar) [inline Tests/PackageLoading/ConventionTests.swift]: 👍

File: Tests/PackageLoading/ConventionTests.swift [hunk 4]
```diff
@@ -440,6 +440,95 @@ class ConventionTests: XCTestCase {
...
+        // Reference a target which doesn't exist.
+        var fs = InMemoryFileSystem()
+        try fs.createEmptyFiles("/Foo.swift")
+        var package = PackageDescription.Package(name: "pkg", targets: [Target(name: "Random")])
+        PackageBuilderTester(package, in: fs) { result in
+            result.checkDiagnostic("these referenced modules could not be found: Random fix: reference only valid modules")
+        }
+
+        // Executable as dependency.
+        // FIXME: maybe should support this and condiser it as build order dependency.
```
Comment (ddunbar) [inline Tests/PackageLoading/ConventionTests.swift]: I think we should.

File: Tests/PackageLoading/ConventionTests.swift [hunk 5]
```diff
@@ -552,6 +641,36 @@ class ConventionTests: XCTestCase {
         }
     }
 
+    func testNoSourcesInModule() throws {
+        let fs = InMemoryFileSystem()
+        try fs.createDirectory(AbsolutePath("/Sources/Module"), recursive: true)
+
+        PackageBuilderTester("MyPackage", in: fs) { result in
+            result.checkDiagnostic("the module at /Sources/Module does not contain any source files fix: either remove the module folder, or add a source file to the module")
```
Comment (ddunbar) [inline Tests/PackageLoading/ConventionTests.swift]: Also OT, but lately I've wondered whether this should be a warning not an error.

I find it really annoying when stashing sources for a new module, but stray editor temp files make the directory stay around and I have to clean it before SwiftPM is willing to build.

File: Tests/PackageLoading/ConventionTests.swift [hunk 6]
```diff
@@ -552,6 +641,36 @@ class ConventionTests: XCTestCase {
...
+        }
+    }
+
+    func testExcludes() throws {
```
Comment (ddunbar) [inline Tests/PackageLoading/ConventionTests.swift]: 👏

File: Tests/PackageLoading/ConventionTests.swift (hunk 1 above)
Comment (aciidb0mb3r) [inline Tests/PackageLoading/ConventionTests.swift]: hmm right

File: Sources/PackageLoading/PackageBuilder.swift [hunk 7]
```diff
@@ -275,13 +275,7 @@ public struct PackageBuilder {
...
         } else {
             modules = try maybeModules.map { path in
-                let name: String
-                if path == srcroot {
-                    name = manifest.name
-                } else {
-                    name = path.basename
-                }
-                return try modulify(path, name: name, isTest: false)
+                try modulify(path, name: path.basename, isTest: false)
```
Comment (ddunbar) [inline Sources/PackageLoading/PackageBuilder.swift]: Thanks, I wondered about this too but didn't stare at it long enough to figure out what it was doing.

File: Sources/PackageModel/Product.swift [hunk 8]
```diff
@@ -70,3 +70,18 @@ extension Product: CustomStringConvertible {
...
+    case (.Executable, .Executable):
+        return true
+    case (.Library(.Static), .Library(.Static)):
+        return true
+    case (.Library(.Dynamic), .Library(.Dynamic)):
+        return true
+    case (.Test, .Test):
+        return true
+    default: return false
+    }
```
Comment (ddunbar) [inline Sources/PackageModel/Product.swift]: BTW, one trick for writing these functions is:
```
switch (lhs, hrs) {
case (.Exectuable, .Executable):
  return true
case (.Executable, _):
  return false
...
}
```
which is longer, but lets you avoid the `default` case which means that this code will fail if it isn't updated when the enum is extended.

File: Tests/PackageLoading/ConventionTests.swift (hunk 2 above)
Comment (aciidb0mb3r) [inline Tests/PackageLoading/ConventionTests.swift]: I think we should, I like how you can quickly create a package. However we do have a proper init now and this becomes untidy once your package grows.
maybe @czechboy0 has some stats about how much this convention is used?

File: Tests/PackageLoading/ConventionTests.swift [hunk 9]
```diff
@@ -737,11 +776,12 @@ private extension FileSystem {
 ///     - package: PackageDescription instance to use for loading this package.
 ///     - path: Directory where the package is located.
 ///     - in: FileSystem in which the package should be loaded from.
+///     - products: List of products in the package.
 ///     - warningStream: OutputByteStream to be passed to package builder.
 ///
 /// - Throws: ModuleError, ProductError
-private func loadPackage(_ package: PackageDescription.Package, path: AbsolutePath, in fs: FileSystem, warningStream: OutputByteStream) throws -> PackageModel.Package {
-    let manifest = Manifest(path: path.appending(component: Manifest.filename), url: "", package: package, products: [], version: nil)
+private func loadPackage(_ package: PackageDescription.Package, path: AbsolutePath, in fs: FileSystem, products: [PackageDescription.Product], warningStream: OutputByteStream) throws -> PackageModel.Package {
```
Comment (ddunbar) [inline Tests/PackageLoading/ConventionTests.swift]: Sigh. The products should have always been a part of the package. We might want to make them that way now even if it isn't what the API is like currently.

File: Sources/PackageModel/Product.swift (hunk 8 above)
Comment (aciidb0mb3r) [inline Sources/PackageModel/Product.swift]: Ah that makes sense, will update

File: Tests/PackageLoading/ConventionTests.swift [hunk 10]
```diff
@@ -827,6 +867,30 @@ final class PackageBuilderTester {
...
+
+    final class ProductResult {
+        private let product: PackageModel.Product
+
+        init(_ product: PackageModel.Product) {
+            self.product = product
+        }
+
+        func check(type: PackageDescription.ProductType, modules: [String], file: StaticString = #file, line: UInt = #line) {
+            // FIXME: Make ProductType equatable.
```
Comment (ddunbar) [inline Tests/PackageLoading/ConventionTests.swift]: You fixed this. :)

**Comment by ddunbar** [comment]
Super-fantastic!!!! `pow(💯, 💯)`

File: Tests/PackageLoading/ConventionTests.swift (hunk 10 above)
Comment (aciidb0mb3r) [inline Tests/PackageLoading/ConventionTests.swift]: Oops 😄

File: Tests/PackageLoading/ConventionTests.swift (hunk 9 above)
Comment (aciidb0mb3r) [inline Tests/PackageLoading/ConventionTests.swift]: Yeah I feet weird writing this. I'll try refactoring that later

File: Tests/PackageLoading/ConventionTests.swift (hunk 5 above)
Comment (aciidb0mb3r) [inline Tests/PackageLoading/ConventionTests.swift]: I wonder why it wasn't supported initially, any downsides?

**Pull request merged**